presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedHistogram.java

```diff
@@ -0,0 +1,205 @@
+
+    private int maxFill;
+    private int mask;
+
+    private Type type;
+
+    private BlockBuilder values;
+    private IntBigArray hashPositions;
+    private final LongBigArray counts;
+
+    public TypedHistogram(Type type, int expectedSize)
+    {
+        this.type = type;
+
+        checkArgument(expectedSize > 0, "expectedSize must be greater than zero");
+
+        int hashSize = arraySize(expectedSize, FILL_RATIO);
+
+        maxFill = calculateMaxFill(hashSize);
+        mask = hashSize - 1;
+        values = this.type.createBlockBuilder(new BlockBuilderStatus(), hashSize, expectedValueSize(type, hashSize));
+        hashPositions = new IntBigArray(-1);
+        hashPositions.ensureCapacity(hashSize);
+        counts = new LongBigArray();
+        counts.ensureCapacity(hashSize);
+    }
+
+    public TypedHistogram(Slice serialized, Type type, int expectedSize)
+    {
+        this(type, expectedSize);
+        checkNotNull(serialized, "serialized is null");
+        deserialize(serialized);
+    }
+
+    public long getEstimatedSize()
+    {
+        return values.getRetainedSizeInBytes() +
+                counts.sizeOf() +
+                hashPositions.sizeOf();
+    }
+
+    private Block getValues()
+    {
+        return values.build();
+    }
+
+    private LongBigArray getCounts()
+    {
+        return counts;
+    }
+
+    private void deserialize(Slice serialized)
+    {
+        Block block = TypeUtils.readStructuralBlock(serialized);
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            add(i, block, block.getLong(i + 1, 0));
+        }
+    }
+
+    public Slice serialize()
+    {
+        Block valuesBlock = values.build();
+        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), valuesBlock.getSizeInBytes() + (int) counts.sizeOf());
+        for (int i = 0; i < valuesBlock.getPositionCount(); i++) {
+            type.appendTo(valuesBlock, i, blockBuilder);
+            BIGINT.writeLong(blockBuilder, counts.get(i));
+        }
+        return buildStructuralSlice(blockBuilder);
+    }
+
+    public void addAll(TypedHistogram other)
+    {
+        Block otherValues = other.getValues();
+        LongBigArray otherCounts = other.getCounts();
+        for (int i = 0; i < otherValues.getPositionCount(); i++) {
+            long count = otherCounts.get(i);
+            if (count > 0) {
+                add(i, otherValues, count);
+            }
+        }
+    }
+
+    public void add(int position, Block block, long count)
+    {
+        int hashPosition = getHashPosition(TypeUtils.hashPosition(type, block, position), mask);
+
+        // look for an empty slot or a slot containing this key
+        while (true) {
+            if (hashPositions.get(hashPosition) == -1) {
+                break;
+            }
+
+            if (block.equals(position, 0, values, hashPositions.get(hashPosition), 0, block.getLength(position))) {
```

Comment:
`block.equals` compares bytes. You need to compare the values (https://prestodb.io/docs/current/develop/types.html), so you need to use `type.equalTo`